src/Microsoft.VisualStudio.LanguageServices.Razor/ProjectSystem/RazorProjectHost.cs

```diff
@@ -0,0 +1,182 @@
+        }
+
+
+        [AppliesTo("DotNetCoreRazor")]
+        [ProjectAutoLoad(ProjectLoadCheckpoint.ProjectFactoryCompleted)]
+        public Task LoadAsync()
+        {
+            return InitializeAsync();
+        }
+
+        // Should only be called from the UI thread.
+        private ProjectSnapshotManagerBase GetProjectManager()
+        {
+            _commonServices.ThreadingService.VerifyOnUIThread();
+
+            if (_projectManager == null)
+            {
+                _projectManager = (ProjectSnapshotManagerBase)_workspace.Services.GetLanguageServices(RazorLanguage.Name).GetRequiredService<ProjectSnapshotManager>();
+            }
+
+            return _projectManager;
+        }
+
+        protected override Task InitializeCoreAsync(CancellationToken cancellationToken)
+        {
+            // Don't try to evaluate any properties here since the project is still loading and we require access
+            // to the UI thread to push our updates.
+            //
+            // Just subscribe and handle the notification later.
+            var receiver = new ActionBlock<IProjectVersionedValue<IProjectSubscriptionUpdate>>(OnProjectChanged);
+            _subscription = _commonServices.ActiveConfiguredProjectSubscription.JointRuleSource.SourceBlock.LinkTo(
+                receiver,
+                initialDataAsNew: true,
+                suppressVersionOnlyUpdates: true,
+                ruleNames: new string[] { RazorGeneral.SchemaName, });
+
+            return Task.CompletedTask;
+        }
+
+        protected override async Task DisposeCoreAsync(bool initialized)
+        {
+            _subscription.Dispose();
+
+            await ExecuteWithLock(async () =>
+            {
+                if (_current != null)
+                {
+                    await _commonServices.ThreadingService.SwitchToUIThread();
+
+                    var projectManager = GetProjectManager();
+                    projectManager.HostProjectRemoved(_current);
+                    _current = null;
+                }
+            });
+        }
+
+        // Async for tests
+        internal async Task OnProjectChanged(IProjectVersionedValue<IProjectSubscriptionUpdate> update)
+        {
+            await ExecuteWithLock(async () =>
+            {
+                if (IsDisposing || IsDisposed)
+                {
+                    return;
+                }
+
+                await _commonServices.ThreadingService.SwitchToUIThread();
+                var projectManager = GetProjectManager();
+
+                var razorProperties = await _commonServices.ActiveConfiguredProjectRazorProperties.GetRazorGeneralPropertiesAsync().ConfigureAwait(false);
+                var languageVersion = await razorProperties.RazorLangVersion.GetEvaluatedValueAtEndAsync().ConfigureAwait(false);
+
+                if (string.IsNullOrEmpty(languageVersion))
+                {
+                    languageVersion = "2.1";
+                }
```

Comment:
We don't currently handle the case where a project file hasn't set the language version